matsim/src/main/java/org/matsim/core/network/DisallowedNextLinks.java

```diff
@@ -0,0 +1,105 @@
+package org.matsim.core.network;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.network.Link;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Class to store disallowed next links, e.g. to model turn restrictions.
+ * 
+ * @author hrewald, mrieser
+ */
+public class DisallowedNextLinks {
+
+	// Actually, it could be Map<String, Set<List<Id<Link>>>>, as the order of the
+	// next links sequences does not matter. However, we choose to store them in a
+	// list in favor of a smaller memory footprint.
+	private final Map<String, List<List<Id<Link>>>> linkIdSequencesMap = new HashMap<>();
+
+	/**
+	 * Add a sequence of subsequent links to be disallowed from the current link.
+	 * 
+	 * @param mode
+	 * @param linkSequence sequence of links that shall not be passed after passing
+	 *                     the link where this object is attached
+	 * @return true, if linkSequence was actually added
+	 */
+	public boolean addDisallowedLinkSequence(String mode, List<Id<Link>> linkSequence) {
+		List<List<Id<Link>>> linkSequences = this.linkIdSequencesMap.computeIfAbsent(mode, m -> new ArrayList<>());
+
+		boolean result = false;
+		// prevent adding empty/duplicate link id sequences, or duplicate link ids
+		if (!linkSequence.isEmpty() && new HashSet<>(linkSequence).size() == linkSequence.size()
+				&& !linkSequences.contains(linkSequence)) {
+			result = linkSequences.add(ImmutableList.copyOf(linkSequence));
+			if (result) { // sorting is required for a.equals(b) <=> a.hashCode() == b.hashCode()
+				Collections.sort(linkSequences, (l, r) -> l.toString().compareTo(r.toString()));
+			}
+		}
+		return result;
+	}
+
+	public List<List<Id<Link>>> getDisallowedLinkSequences(String mode) {
+		List<List<Id<Link>>> sequences = this.linkIdSequencesMap.get(mode);
+		return sequences != null ? Collections.unmodifiableList(sequences) : Collections.emptyList();
+	}
+
+	@Nullable
+	public List<List<Id<Link>>> removeDisallowedLinkSequences(String mode) {
+		return this.linkIdSequencesMap.remove(mode);
+	}
+
+	public Map<String, List<List<Id<Link>>>> getAsMap() {
+		return this.linkIdSequencesMap.entrySet().stream()
+				.collect(Collectors.toMap(Entry::getKey, e -> Collections.unmodifiableList(e.getValue())));
+	}
+
+	public void clear() {
+		this.linkIdSequencesMap.clear();
+	}
+
+	public boolean isEmpty() {
+		return this.linkIdSequencesMap.isEmpty();
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof DisallowedNextLinks dnl) {
```

Comment:
Early return if not `instanceof`?

Reply:
I don't see how this would be simpler, maybe you can help me out?
I only could prepend these lines, but it would be too redundant, I think:
```java
if (!(obj instanceof DisallowedNextLinks)) {
	return false;
}
```

Reply:
Yes, this is what I was thinking about. This would reduce the level of nesting by 1 in the following `if` statements.

But it's just a hint, you can ignore it :-)

Reply:
You are right, it was possible to remove that one nesting, thanks!